actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -6,30 +6,72 @@ module ActionDispatch
   class ServerTiming
     SERVER_TIMING_HEADER = "Server-Timing"
 
+    class Subscriber # :nodoc:
+      include Singleton
+      KEY = :action_dispatch_server_timing_events
+
+      def initialize
+        @mutex = Mutex.new
+      end
+
+      def call(event)
+        if events = ActiveSupport::IsolatedExecutionState[KEY]
+          events << event
+        end
+      end
+
+      def collect_events
+        events = []
+        ActiveSupport::IsolatedExecutionState[KEY] = events
+        yield
+        events
+      ensure
+        ActiveSupport::IsolatedExecutionState.delete(KEY)
+      end
+
+      def ensure_subscribed
+        @mutex.synchronize do
```

Comment:
I think we could use double checked locking to avoid synchronizing that mutex on every request.

Reply:
I'd prefer readability here, and I think it's a safe bet that anyone enabling this (presumably only in development) isn't too concerned with performance.

I could see revisiting this if we make this more production-appropriate (sampling a small set of useful topics), but that would likely look different from this.